### PR TITLE
Add StreamingFeatureBufferer class for real-life streaming decoding

### PR DESCRIPTION
**Pull request by tango4j**
Signed-off-by: Taejin Park <tango4j@gmail.com>

# What does this PR do ?

Add StreamingFeatureBufferer class for real-life streaming decoding

**Collection**: [Note which collection this PR will affect]
ASR

# Changelog 

Added StreamingFeatureBufferer() in streaming_utils.py.

# Usage

Will add usage when the example code is ready and pushed.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in ASR

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/utils/streaming_utils.py [hunk 1]
```diff
@@ -343,6 +343,129 @@ def inplace_buffer_merge(buffer, data, timesteps, model):
...
+        self.buffer = np.ones([self.n_feat, total_buffer_len], dtype=np.float32) * self.ZERO_LEVEL_SPEC_DB_VAL
+        self.feature_chunk_len = int(chunk_size / timestep_duration)
+
+        self.feature_buffer_len = total_buffer_len
+
+        self.feature_buffer = (
+            np.ones([self.n_feat, self.feature_buffer_len], dtype=np.float32) * self.ZERO_LEVEL_SPEC_DB_VAL
+        )
+        self.reset()
+        self.buffered_len = 0
```
Comment (VahidooX) [inline nemo/collections/asr/parts/utils/streaming_utils.py:387]: Has buffered_len being used in the code?

File: nemo/collections/asr/parts/utils/streaming_utils.py [hunk 2]
```diff
@@ -343,6 +343,129 @@ def inplace_buffer_merge(buffer, data, timesteps, model):
...
+        """
+        Add an extracted feature to `feature_buffer`
+        """
+        self.feature_buffer[:, : -self.feature_chunk_len] = self.feature_buffer[:, self.feature_chunk_len :]
+        self.feature_buffer[:, -self.feature_chunk_len :] = feat_chunk
+
+    def get_raw_feature_buffer(self):
+        return self.feature_buffer
+
+    def get_normalized_feature_buffer(self):
```
Comment (VahidooX) [inline nemo/collections/asr/parts/utils/streaming_utils.py:429]: Isn't it better to use normalize_batch method here?
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/preprocessing/features.py#L59

And pass the normalize_type which is extracted already from the asr_model in the init
.

File: nemo/collections/asr/parts/utils/streaming_utils.py [hunk 3]
```diff
@@ -343,6 +343,129 @@ def inplace_buffer_merge(buffer, data, timesteps, model):
...
+                Size of the look-back length for middle chunk
+        '''
+
+        self.NORM_CONSTANT = 1e-5
+        self.ZERO_LEVEL_SPEC_DB_VAL = -16.635  # Log-Melspectrogram value for zero signal
+        self.asr_model = asr_model
+        self.sr = asr_model._cfg.sample_rate
+        self.chunk_size = chunk_size
+        self.n_chunk_look_back = int(chunk_look_back * self.sr)
+        timestep_duration = asr_model._cfg.preprocessor.window_stride
```
Comment (VahidooX) [inline nemo/collections/asr/parts/utils/streaming_utils.py:372]: I think it is better to use cfg instead of _cfg. @titu1994 Is it right?

File: nemo/collections/asr/parts/utils/streaming_utils.py [hunk 4]
```diff
@@ -343,6 +343,129 @@ def inplace_buffer_merge(buffer, data, timesteps, model):
...
 
 
+class StreamingFeatureBufferer:
+    """
+    Class to append each feature frame to a buffer and return an array of buffers.
+    This class is designed to perform a real-life streaming decoding where only a single chunk
+    is provided at each step of a streaming pipeline.
+    """
+
+    def __init__(self, asr_model, chunk_size=1.6, buffer_size=5.0, chunk_look_back=0.01):
```
Comment (VahidooX) [inline nemo/collections/asr/parts/utils/streaming_utils.py:353]: Maybe not having default value for chunk_size and buffer_size as user need to specify them? Your call.

**Review by VahidooX — CHANGES_REQUESTED** [review]
LGTM! Just minor comments.

File: nemo/collections/asr/parts/utils/streaming_utils.py (hunk 2 above)
Comment (jbalam-nv) [reply nemo/collections/asr/parts/utils/streaming_utils.py:429]: This makes sense. This class is based on another older class we used for streaming simulation. @tango4j I will clean up both the classes later, you can leave this as is for now.

File: nemo/collections/asr/parts/utils/streaming_utils.py (hunk 4 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/streaming_utils.py:353]: removed default values

File: nemo/collections/asr/parts/utils/streaming_utils.py (hunk 3 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/streaming_utils.py:372]: changed it to cfg

File: nemo/collections/asr/parts/utils/streaming_utils.py (hunk 1 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/streaming_utils.py:387]: removed redundant declaration

File: nemo/collections/asr/parts/utils/streaming_utils.py (hunk 2 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/streaming_utils.py:429]: I have talked to @jbalam-nv and I converted the whole class to torch.tensor based + used `normalize_batch()` function for consistency.

**Review by VahidooX — APPROVED** [review]
